add cohen kappa

KickItLikeShika opened the pull request:
Fixes #1673

Added Cohen's Kappa in `ignite.contrib.metrics`. Implemented 'CohenKappa' in `cohen_kappa.py` file. And added the tests in `test_cohen_kappa.py`.

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

KickItLikeShika commented:
@vfdev-5. Putting try/except in the `__init__` didn't work. I have simplified the implementation a lot more.

vfdev-5 commented:
Hey @KickItLikeShika , what's the issue with try/except in the `__init__` ?

As for other updates, please, take a look at a similar PR here and the comments about the documentation : https://github.com/pytorch/ignite/pull/1682

As discussed in your previous closed PR, please use a single PR to bring all the updates according to the review. Otherwise, we wont be able to consider them.

KickItLikeShika commented:
@vfdev-5. I have fixed it. Is there anything else to edit?

vfdev-5 commented on ignite/contrib/metrics/cohen_kappa.py:
```diff
@@ -0,0 +1,63 @@
+    generated/sklearn.metrics.cohen_kappa_score.html>`_ .
+
+    Args:
+        output_transform (callable, optional): a callable that is used to transform the
```
Please, remove the type hints
```suggestion
        output_transform: a callable that is used to transform the
```
and same below.

vfdev-5 commented on ignite/contrib/metrics/cohen_kappa.py:
```diff
@@ -0,0 +1,63 @@
+
+    def get_cohen_kappa_fn(self) -> Callable[[torch.Tensor, torch.Tensor], float]:
+
+        try:
```
What is not working with moving try/except to `__init__` ?

vfdev-5 commented on ignite/contrib/metrics/cohen_kappa.py:
```diff
@@ -0,0 +1,63 @@
+
+    """
+
+    def __init__(self, output_transform: Callable = lambda x: x, weights: str = "", check_compute_fn: bool = False):
```
```suggestion
    def __init__(self, output_transform: Callable = lambda x: x, weights: Optional[str] = None, check_compute_fn: bool = False):
```

vfdev-5 commented on ignite/contrib/metrics/cohen_kappa.py:
```diff
@@ -0,0 +1,63 @@
+
+    def __init__(self, output_transform: Callable = lambda x: x, weights: str = "", check_compute_fn: bool = False):
+        # initalize weights
+        if weights == "":
+            self.weights = None
+        else:
+            self.weights = weights
```
```suggestion
        self.weights = weights
```

KickItLikeShika replied:
This doesn't pass the CI, because of conversion between `NoneType`, `str`. The weights takes a `None` or `linear` of `quadratic`. So i made this hack to pass the CI

vfdev-5 commented on tests/ignite/contrib/metrics/test_cohen_kappa.py:
```diff
@@ -0,0 +1,148 @@
+    data = list(range(size // batch_size))
+    ck_value = engine.run(data, max_epochs=1).metrics["cohen_kappa"]
+
+    assert np.array_equal(ck_value, ck_value_sk)
```
We need to add distributed tests as well. See https://github.com/pytorch/ignite/blob/master/tests/ignite/contrib/metrics/regression/test_canberra_metric.py#L99

KickItLikeShika replied:
We are using sklearn just like average precision, can we change the device? I guess it's not possible, the same thing for `roc curve`, `average precision`, `precision recall curve`.

vfdev-5 replied:
Yes, you are right, we are missing an argument in the constructor, see https://github.com/pytorch/ignite/blob/d38c8bcb1794b85c3525c8cc1862f7e520d7862c/ignite/metrics/epoch_metric.py#L44

The point here is that we have to start providing distributed support for all metrics (see https://github.com/pytorch/ignite/issues/1284). EpochMetric, base class for this metric, now supports distributed configuration and performs necessary collective op (all_gather) to fetch all the data from each participating proc and executes `compute_fn` on the rank 0 passing gathered data (https://github.com/pytorch/ignite/blob/d38c8bcb1794b85c3525c8cc1862f7e520d7862c/ignite/metrics/epoch_metric.py#L123-L147).
Currently, we are assuming that device is CPU and gathered tensors are also on CPU. And yes, you are correct, that if user specifies `device` as CUDA, calling `y_targets.numpy()` will fail. We have to fix that and as I said add distributed tests on cpu and gpu.

KickItLikeShika replied:
Okay i will add the device arg for now and check if the user specified CUDA then i will raise an exception that it's not supported yet.

vfdev-5 replied:
You can simply call `y_targets.cpu().numpy()` for other device than cpu. No need to raise an error.

vfdev-5 reviewed (commented):
@KickItLikeShika few more comments to address, please.

vfdev-5 commented on ignite/contrib/metrics/cohen_kappa.py:
```diff
@@ -0,0 +1,64 @@
+        self, output_transform: Callable = lambda x: x, weights: Optional[str] = None, check_compute_fn: bool = False
+    ):
+
+        try:
```
By the way, could you please open an issue to improve AveragePrecision and others metrics to have a similar logic there: we ensure that sklearn is present in the constructor and import necessary sklearn function inside their own `compute_fn`. Please, also mention that we have to add `device` as argument as well.

KickItLikeShika replied:
Sure!

vfdev-5 replied:
A small suggestion on the github PR's comment discussion: please, resolve the conversation if it does not require any further discussion and required changes were done.

KickItLikeShika commented:
@vfdev-5. I have updated everything as you said and added the test. But please look at these CI errors, it works fine locally! and look at the values, it so close! That's weird!

vfdev-5 commented:
@KickItLikeShika sounds great for the update ! Yes, it can happen that values could differ with a tiny eps. When you say that it works locally, how do you run the tests ?

KickItLikeShika commented:
`pytest test_cohen_kappa.py`, that's how i run them.

vfdev-5 commented on ignite/contrib/metrics/cohen_kappa.py:
```diff
@@ -0,0 +1,72 @@
+            <https://scikit-learn.org/stable/modules/generated/sklearn.metrics.cohen_kappa_score.html>`_
+            is run on the first batch of data to ensure there are
+            no issues. User will be warned in case there are any issues computing the function.
+        device (str or torch.device, optional): optional device specification for internal storage.
```
```suggestion
        device: optional device specification for internal storage.
```

vfdev-5 commented:
If you'd like to replicate distributed tests on CPU, you have to replicate the command from `tests/run_cpu_tests.sh`:
```
export WORLD_SIZE=2
CUDA_VISIBLE_DEVICES="" pytest --dist=each --tx $WORLD_SIZE*popen//python=python tests -m distributed -vvv
```
and make sure to install `pytest-xdist`.

KickItLikeShika commented:
Okay thank you! Is there anything else to work on in this pull request?

vfdev-5 commented on ignite/contrib/metrics/cohen_kappa.py:
```diff
@@ -0,0 +1,72 @@
+            :class:`~ignite.engine.engine.Engine`'s ``process_function``'s output into the
+            form expected by the metric. This can be useful if, for example, you have a multi-output model and
+            you want to compute the metric with respect to one of the outputs.
+        weights: a string is used to define the type of Cohen's Kappa whether Non-Weighted or Linear or Quadratic
+            (default: "")
```
```suggestion
        weights: a string is used to define the type of Cohen's Kappa whether Non-Weighted or Linear 
            or Quadratic. Default, None.
```

vfdev-5 commented on ignite/contrib/metrics/cohen_kappa.py:
```diff
@@ -0,0 +1,72 @@
+            raise RuntimeError("This contrib module requires sklearn to be installed.")
+
+        # initalize weights
+        self.weights = weights
```
Shouldn't we check provided `weights` values ? Like if user mispell the name etc.

KickItLikeShika replied:
But that's already provided in sklearn 
`
ValueError: Unknown kappa weighting type.
`

vfdev-5 replied:
Yes, but user will get this error in the end of the epoch (after a certain time) if not using `check_compute_fn=True` which is by default `False`.

KickItLikeShika replied:
Okay i will add the checks and the `pytest.approx`

vfdev-5 replied:
And I forgot to mention that we have to run a check for that to ensure better codecov. Search for `with pytest.raises(...):` in the codebase to see how to do that.

KickItLikeShika replied:
Well it seems it's the only error in the CI we got not, i will do it.

KickItLikeShika replied:
@vfdev-5, all updated.

vfdev-5 commented:
And to fix precision issue, you can use `pytest.approx`, please search for its usage on the repo.

vfdev-5 commented on tests/ignite/contrib/metrics/test_cohen_kappa.py:
```diff
@@ -0,0 +1,299 @@
+        ck._check_shape((torch.randint(0, 2, size=(10, 1)).long(), torch.randint(0, 2, size=(10, 5, 2)).long()))
+
+
+def test_cohen_kappa_non_weighted():
```
Let's try to use `pytest.mark.parametrize` here. For example, take a look here : https://github.com/pytorch/ignite/pull/1683/files#diff-be88f17b02401b784cf4081d7e84bc7470d9beac1ac42294b8a0b1fc731900b3R20

Such that we can refactor all those 3 almost idenitcal functions. Probably, a similar thing can be done for test_integration_cohen_kappa_non_weighted_with_output_transform and its 2 triplets...

KickItLikeShika replied:
I understood how it works. I can use `test.mark.parametrize` to replace all of these function with just 2 or 3. Should i do that?

vfdev-5 replied:
> I understood how it works. I can use test.mark.parametrize to replace all of these function with just 2 or 3. Should i do that?

The idea is to make test code better with minimal code duplication:
3 test functions -> 1 test function with parametrize

- `test_cohen_kappa_*` -> `test_cohen_kappa`
-  `test_integration_cohen_kappa_*_with_output_transform` -> `test_integration_cohen_kappa_with_output_transform`

Sorry, I do not quite get your question...

KickItLikeShika replied:
@vfdev-5, Updated all the tests, and added new tests.

KickItLikeShika replied:
@vfdev-5, It breaks the CI. The issue is the code formatting, `black` and `pycodestyle` don't agree on whitespaces before the columns. Any idea how to fix this?

vfdev-5 replied:
Can you please put a link to where there is a failure between code formatters ?

KickItLikeShika replied:
Aftert running `black test_cohen_kappa.py` the columns in lines 98, 107, 108 get reformatted and whitespaces added before the columns. But then `pycodestyle test_cohen_kappy.py` shows that i should remove the whitespaces before the columns. And that's how the unit tests in the CI fail

KickItLikeShika replied:
@vfdev-5, should i change the logic in `update_fn` in `test_cohen_kappa_all_weights_with_output_transform` for avoid these CI errors?

vfdev-5 replied:
@KickItLikeShika I do not quite understand your problem with formatting here. I updated your branch to the master to see where it is failing.

vfdev-5 replied:
Looks like it is OK.

KickItLikeShika replied:
Well! i will commit the latest updates now with and delete the tests you told my about

vfdev-5 reviewed (commented):
@KickItLikeShika thanks for a thourough testing !
I have a remark about duplicating test functions and usage of parametrize.

vfdev-5 commented:
@KickItLikeShika as a suggestion, try to enable Github Actions on your fork such that our "Format python code / code-style" could commit code format fixed files to your fork and avoid manual fixing.

vfdev-5 commented on tests/ignite/contrib/metrics/test_cohen_kappa.py:
```diff
@@ -0,0 +1,208 @@
+    with pytest.raises(ValueError):
+        CohenKappa(weights="dd")
+
+    with pytest.raises(ValueError):
+        CohenKappa(weights="ss")
+
+    with pytest.raises(ValueError):
+        CohenKappa(weights="l")
+
+    with pytest.raises(ValueError):
+        CohenKappa(weights="q")
+
+    with pytest.raises(ValueError):
+        CohenKappa(weights="z")
```
I think all those tests can be removed

KickItLikeShika replied:
Okay done

vfdev-5 replied:
@KickItLikeShika remove the lines from 75 - 82.

vfdev-5 commented on tests/ignite/contrib/metrics/test_cohen_kappa.py:
```diff
@@ -0,0 +1,208 @@
+def test_input_types():
+    ck = CohenKappa()
+
+    with pytest.raises(ValueError):
```
Please, use `with pytest.raises(ValueError, match=r"...")` to match exact error. You can match first 2-3 significant words. Same for all `with pytest.raises(` below. Above `with pytest.raises(NotComputableError)` is OK.

KickItLikeShika replied:
That's done too. Check the recent commit.

vfdev-5 replied:
@KickItLikeShika, sorry, I do not understand why did you remove all the tests like 
```python
    with pytest.raises(ValueError):
        CohenKappa(weights="ss")
```
I think I asked you to remove redundent ones only and not all, right. Am I missing something ?

vfdev-5 replied:
Anyway, thanks for the update, let's restore removed tests **only relevant ones** and it will be good to go.

KickItLikeShika replied:
I thought you told me to remove all of them! Sorry i will add them back

vfdev-5 reviewed (approved):
Thanks for the PR @KickItLikeShika ! Looks good now 👍

KickItLikeShika commented:
@vfdev-5, thank you! I will work now on #1695

vfdev-5 commented:
Actually, I was thinking that TPU test failure is unrelated but it is related:
```
tests/ignite/contrib/metrics/test_cohen_kappa.py::test_distrib_single_device_xla PASSED [  3%]
tests/ignite/contrib/metrics/test_cohen_kappa.py::test_distrib_xla_nprocs SKIPPED [  5%]
tests/ignite/contrib/metrics/regression/test_canberra_metric.py::test_distrib_single_device_xla FAILED [  6%]
```
as it does not fail on other PR, nor on master.

KickItLikeShika commented:
Do you have any idea what might be the problem here?

KickItLikeShika commented:
@vfdev-5, Do you think i have to edit anything before i change the rest of the metrics to this similar implementation?

vfdev-5 commented:
> @vfdev-5, Do you think i have to edit anything before i change the rest of the metrics to this similar implementation?

Well, the test with XLA should be fixed. As far as I can see from https://github.com/pytorch/ignite/pull/1238/commits/61e50414d3b5d5dd315592e0b0ec4e0569b78fec#diff-54ddf6f7687e40789e7b5180b452037c40cb021140ae45ed5297d5698a6a8dbeR164 we disabled accumulation on XLA devices (unfortunately without any explanation). We have to invetigate this more in details.
Temporarily, we can comment out XLA tests and I'll merge the PR.

KickItLikeShika commented:
@vfdev-5 then i will start improving the other metrics the same way and add the tests

vfdev-5 commented:
@KickItLikeShika  finally get it done 🎉 Congrats :)

KickItLikeShika commented:
@vfdev-5, Thank you!!
